arch/arm/core/aarch32/swap_helper.S

```diff
@@ -71,6 +71,10 @@ SECTION_FUNC(TEXT, z_arm_pendsv)
     ldr r1, =_kernel
     ldr r2, [r1, #_kernel_offset_to_current]
 
+#ifdef CONFIG_ARM_SWAP_EXC_RETURN
+    str lr, [r2, #_thread_offset_to_exc_return]
```

Comment:
I think you are not always storing the EXC_RETURN here, it might have changed due to the call to z_thread_mark_switched_out